### PR TITLE
(MODULES-6895) Refactor provider code to Trigger

### PR DESCRIPTION
**Pull request by Iristyle**
Completely overhaul Trigger manipulation code, removing it from the provider instance and into the Trigger class.

See individual commits for more details, but note that there are a number of new classes:

* Trigger::V1::Day
* Trigger::V1::Days
* Trigger::V1::Month
* Trigger::V1::Occurrence
* Trigger::V1::Flag

### CODE REVIEW

File: lib/puppet/provider/scheduled_task/taskscheduler_api2.rb [hunk 1]
```diff
@@ -313,7 +313,7 @@ def translate_hash_to_trigger(puppet_trigger)
       }
 
       trigger['type']['days_of_week'] = if puppet_trigger['day_of_week']
-                                          PuppetX::PuppetLabs::ScheduledTask::Trigger::V1::Day.bitfield_from_days_of_week(puppet_trigger['day_of_week'])
+                                          PuppetX::PuppetLabs::ScheduledTask::Trigger::V1::Day.names_to_bitfield(puppet_trigger['day_of_week'])
```
Comment (michaeltlombardi) [inline lib/puppet/provider/scheduled_task/taskscheduler_api2.rb]: Much easier to follow with the renaming

Comment (glennsarti) [reply lib/puppet/provider/scheduled_task/taskscheduler_api2.rb]: nit: Typically that's a bitmask not a bitfield.  I don't think it's worth changing though, particularly if it's used everywhere

File: lib/puppet_x/puppetlabs/scheduled_task/trigger.rb [hunk 2]
```diff
@@ -106,6 +106,39 @@ class V1
...
+
+      DAY_CONST_MAP = {
+        'sun'   => TASK_SUNDAY,
+        'mon'   => TASK_MONDAY,
+        'tues'  => TASK_TUESDAY,
+        'wed'   => TASK_WEDNESDAY,
+        'thurs' => TASK_THURSDAY,
+        'fri'   => TASK_FRIDAY,
+        'sat'   => TASK_SATURDAY,
+      }
```
Comment (glennsarti) [inline lib/puppet_x/puppetlabs/scheduled_task/trigger.rb]: nit: Should this be frozen?

File: lib/puppet_x/puppetlabs/scheduled_task/trigger.rb [hunk 3]
```diff
@@ -148,6 +148,72 @@ def self.bitfield_to_names(bitfield)
...
+        4  => TASK_APRIL,
+        5  => TASK_MAY,
+        6  => TASK_JUNE,
+        7  => TASK_JULY,
+        8  => TASK_AUGUST,
+        9  => TASK_SEPTEMBER,
+        10 => TASK_OCTOBER,
+        11 => TASK_NOVEMBER,
+        12 => TASK_DECEMBER,
+      }
```
Comment (glennsarti) [inline lib/puppet_x/puppetlabs/scheduled_task/trigger.rb]: nit: Should this be frozen?

File: lib/puppet/provider/scheduled_task/taskscheduler_api2.rb [hunk 4]
```diff
@@ -95,11 +95,11 @@ def trigger
         puppet_trigger['day_of_week'] = PuppetX::PuppetLabs::ScheduledTask::Trigger::V1::Day.bitfield_to_names(trigger['type']['days_of_week'])
       when Win32::TaskScheduler::TASK_TIME_TRIGGER_MONTHLYDATE
         puppet_trigger['schedule'] = 'monthly'
-        puppet_trigger['months']   = PuppetX::PuppetLabs::ScheduledTask::Trigger::V1::Month.months_from_bitfield(trigger['type']['months'])
+        puppet_trigger['months']   = PuppetX::PuppetLabs::ScheduledTask::Trigger::V1::Month.bitfield_to_indexes(trigger['type']['months'])
```
Comment (glennsarti) [inline lib/puppet/provider/scheduled_task/taskscheduler_api2.rb]: Not sure how I feel about `indexes`.  Index into what?

File: lib/puppet/provider/scheduled_task/taskscheduler_api2.rb [hunk 5]
```diff
@@ -105,8 +105,8 @@ def trigger
       when Win32::TaskScheduler::TASK_TIME_TRIGGER_ONCE
         puppet_trigger['schedule'] = 'once'
       end
-      puppet_trigger['start_date'] = self.class.normalized_date(trigger['start_year'], trigger['start_month'], trigger['start_day'])
-      puppet_trigger['start_time'] = self.class.normalized_time(trigger['start_hour'], trigger['start_minute'])
+      puppet_trigger['start_date'] = PuppetX::PuppetLabs::ScheduledTask::Trigger::V1.normalized_date(trigger['start_year'], trigger['start_month'], trigger['start_day'])
```
Comment (glennsarti) [inline lib/puppet/provider/scheduled_task/taskscheduler_api2.rb]: Typo in commit message ((MODULES-6895) Move normalized helpers to V1 class): "codd" -> "code"

File: lib/puppet/provider/scheduled_task/taskscheduler_api2.rb (hunk 1 above)
Comment (Iristyle) [reply lib/puppet/provider/scheduled_task/taskscheduler_api2.rb]: Yeah, the data structure is as close to a bitfield as Ruby can manage. But bitmask is certainly the more common term - I'll change it as long as there isn't too much carnage.

File: lib/puppet_x/puppetlabs/scheduled_task/trigger.rb (hunk 3 above)
Comment (Iristyle) [reply lib/puppet_x/puppetlabs/scheduled_task/trigger.rb]: Sure

File: lib/puppet/provider/scheduled_task/taskscheduler_api2.rb (hunk 4 above)
Comment (Iristyle) [reply lib/puppet/provider/scheduled_task/taskscheduler_api2.rb]: They are month indexes .. i.e. 1 through 12. I couldn't readily think of a better name, but open to suggestions if you have any.

Comment (Iristyle) [reply lib/puppet/provider/scheduled_task/taskscheduler_api2.rb]: https://en.wikipedia.org/wiki/Month#Julian_and_Gregorian_calendars doesn't have too much else...

Is `bitmask_to_numbers` any better than `bitmask_to_indexes`?

Comment (michaeltlombardi) [reply lib/puppet/provider/scheduled_task/taskscheduler_api2.rb]: Longer, but possibly `bitmask_to_month_index`?

File: lib/puppet_x/puppetlabs/scheduled_task/trigger.rb [hunk 6]
```diff
@@ -106,6 +106,207 @@ class V1
...
+      TASK_FOURTH_WEEK  = 4
+      TASK_LAST_WEEK    = 5
+
+      CONST_MAP = {
+        'first'  => TASK_FIRST_WEEK,
+        'second' => TASK_SECOND_WEEK,
+        'third'  => TASK_THIRD_WEEK,
+        'fourth' => TASK_FOURTH_WEEK,
+        'last'   => TASK_LAST_WEEK,
+      }
```
Comment (Iristyle) [inline lib/puppet_x/puppetlabs/scheduled_task/trigger.rb]: freeze me ... also, better name than `CONST_MAP`?

Comment (glennsarti) [reply lib/puppet_x/puppetlabs/scheduled_task/trigger.rb]: I was ok with const map as it's internal to the class so easy to reason about.

**Comment by Iristyle** [comment]
CI is failing per https://jenkins-master-prod-1.delivery.puppetlabs.net/view/All/job/forge-windows_puppetlabs-scheduled_task_intn-sys_smoke-adhoc/7/PLATFORM=windows2008r2-64default.a,WORKER_LABEL=beaker/testReport/junit/spec.acceptance/should_create_task_spec/Should_create_a_scheduled_task_Should_create_a_task_if_it_does_not_exist/

Specifically the test at https://github.com/puppetlabs/puppetlabs-scheduled_task/blob/master/spec/acceptance/should_create_task_spec.rb is failing, so we'll need to resolve that one issue prior to merge.

**Comment by glennsarti** [comment]
Error
```
23:07:55   Error: /Stage[main]/Main/Scheduled_task[pl254138]/compatibility: change from 1 to 1 failed: The taskscheduler_api2 provider can not handle attribute compatibility
```

Looks like that's the only test for idempotency too.

**Comment by Iristyle** [comment]
You must have edited your GH comment after initially leaving it @glennsarti? The email I got didn't have any of the info about commits or rebasing, but the initial one was enough to clue me in anyhow. I totally forgot this branch started before the second PR for 6526 landed, which had the acceptance fix. After realizing I needed a rebase, I rekicked adhoc and it has passed at https://jenkins-master-prod-1.delivery.puppetlabs.net/view/All/job/forge-windows_puppetlabs-scheduled_task_i18n-ruby_adhoc/4/


But yeah, the merge would have passed anyhow, but at least we get a quick turnaround on verification this way.


Thanks!

**Pull request merged**